File: PythonSolutions/Neetcode150/linkedlist_problems/2_add-two-numbers.py
from typing import Optional
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = current = ListNode()

        carry = 0

        # Carry is needed as a condition in the while since theres an edge case where both lists are empty
            # Carry may still have a value of 1 when both lists are empty
        while l1 or l2 or carry:
            # If theres a carry number, add to total
            total = carry
            
            # Add the two numbers and traverse them for next iteration
            # *Long addition is done here from left to right rather than right to left due to direction of linked list*
            if l1:
                total += l1.val
                l1 = l1.next
            if l2:
                total += l2.val
                l2 = l2.next

            # Checks if we have to carry any values to next nodes (if total for current column is > 9), e.g. 16 // 10 = 1.6 = 1 
            carry = total // 10
            # Change total using remainder
                # Grabs the remainder which is the 2nd digit of total for current column, e.g. 19 % 10 = 9 
            total = total % 10
            
            current.next = ListNode(total)
            current = current.next
        
        # Return using the reference as we have traversed current node to end of list
        return dummy.next
            

File: TypescriptSolutions/Neetcode150/linked-list_problems/2_add-two-numbers.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

const addTwoNumbers = (l1: ListNode | null, l2: ListNode | null): ListNode | null => {
    let dummy = new ListNode();
    let curr = dummy;

    let carry = 0;
    while (l1 || l2 || carry) {
        let total = carry;

        if (l1) {
            total += l1.val;
            l1 = l1.next;
        }
        if (l2) {
            total += l2.val;
            l2 = l2.next;
        }

        carry = Math.floor(total / 10);
        total = total % 10;

        curr.next = new ListNode(total);
        curr = curr.next;
    }

    return dummy.next;
};